astropy/modeling/functional_models.py

```diff
@@ -1873,73 +1881,16 @@ def _hum2zpf16c(z, s=10.0):
             Z = z[np.where(mask)] + 1.35j
+
+            numer = 0.0
+            for A in AA[::-1]:
+                numer = numer * Z + A
+
             ZZ = Z * Z
-            numer = (
-                (
-                    (
-                        (
-                            (
-                                (
-                                    (
-                                        (
-                                            (
-                                                (
-                                                    (
-                                                        (
-                                                            (
-                                                                (AA[15] * Z + AA[14])
-                                                                * Z
-                                                                + AA[13]
-                                                            )
-                                                            * Z
-                                                            + AA[12]
-                                                        )
-                                                        * Z
-                                                        + AA[11]
-                                                    )
-                                                    * Z
-                                                    + AA[10]
-                                                )
-                                                * Z
-                                                + AA[9]
-                                            )
-                                            * Z
-                                            + AA[8]
-                                        )
-                                        * Z
-                                        + AA[7]
-                                    )
-                                    * Z
-                                    + AA[6]
-                                )
-                                * Z
-                                + AA[5]
-                            )
-                            * Z
-                            + AA[4]
-                        )
-                        * Z
-                        + AA[3]
-                    )
-                    * Z
-                    + AA[2]
-                )
-                * Z
-                + AA[1]
-            ) * Z + AA[0]
-            denom = (
-                (
-                    (
-                        ((((ZZ + bb[14]) * ZZ + bb[12]) * ZZ + bb[10]) * ZZ + bb[8])
-                        * ZZ
-                        + bb[6]
-                    )
-                    * ZZ
-                    + bb[4]
-                )
-                * ZZ
-                + bb[2]
-            ) * ZZ + bb[0]
+            denom = 0.0
+            for b in bb[::-1]:
+                denom = denom * ZZ + b
```

Comment:
Actually shouldn't this just go over the even indices as above?

Reply:
And if so, it seems our test suite isn't catching this?

Reply:
@astrofrog, no the "odd" positions of the array were both all zeros and never accessed. Therefore, I simply removed all those positions and simplified the loop.

Reply:
Also, there is pretty exact input/output testing in the test suite which would be highly unlikely to pass if I made a mistake here.

Reply:
Have you benchmarked this? Making some timing tests on relatively small input arrays (like `lam = np.linspace(-10, 10, 1001)`) for the cases in #11177 I am seeing something like 20 % overhead by just changing to the iterative multiplication above. Should read through the comments in the PR to recall exactly why we chose that form of inline calculation, but I'd strongly recommend to go through at least one more round of review before making actual functional code changes in this context.

> @astrofrog, no the "odd" positions of the array were both all zeros and never accessed. Therefore, I simply removed all those positions and simplified the loop.

That's correct; the declaration of `bb` had just been adopted 1-1 from the original in `cpfX.py`, where it was also used in various other implementations, like using `np.poly1d`, and possibly also for symmetry reasons.
For the

Reply:
> Have you benchmarked this? Making some timing tests on relatively small input arrays (like `lam = np.linspace(-10, 10, 1001)`) for the cases in #11177 I am seeing something like 20 % overhead by just changing to the iterative multiplication above. Should read through the comments in the PR to recall exactly why we chose that form of inline calculation, but I'd strongly recommend to go through at least one more round of review before making actual functional code changes in this context.

I am okay (with reservations) with reverting this on performance grounds. However, I think the current way (before this PR) that the code is written is unreadable. I had to write out by hand the original formulation to figure out what was going on which makes this a rather difficult thing to maintain. Moreover, I dislike the "magic" pre-computed values used by this function in general. If we want to go with reverting this then I would like someone to clearly document the reasoning for the current implementation.

I have not done any benchmarking on this; however, do you know if this 20% is fixed or it relative to problem size. That is, there is a good chance the the overhead cost is relatively fixed as all the cases in #11177 are fairly small (the loop is a fixed iteration size). In any case, if performance is a huge concern for this algorithm, then we should consider moving this algorithm into a C extension and/or contributing it directly to `scipy`?

Reply:
So I just read through #11177 a second time more closely. It is mentioned that `scipy.special.wofz` is almost as fast as this but more accurate. Sure it requires `scipy` but so do lots of other models in astropy. Moreover, since it is more accurate (and presumably better tested and maintained) it should be the default method (when `scipy` is installed). I also would like to go further and say that unless there is a compelling reason to not require `scipy` for the `Voigt1D` model to function, then we should deprecate the `humlicek2` algorithm all together.

Reply:
> I am okay (with reservations) with reverting this on performance grounds. However, I think the current way (before this PR) that the code is written is unreadable. I had to write out by hand the original formulation to figure out what was going on which makes this a rather difficult thing to maintain. Moreover, I dislike the "magic" pre-computed values used by this function in general. If we want to go with reverting this then I would like someone to clearly document the reasoning for the current implementation.

The 4 literature references as well as the link to the original `cpfX.py` (which of course may not be stable, but so far is still online) are all given in the docstrings. The coding is indeed somewhat impenetrable (arguably the original was even more so ;-), but on the other hand, keeping it close to the reference implementation makes it easier to identify when comparing to the source.
But if we decide to keep it that way I also suggest to add a comment that it is an inline formulation of the recursive product `numer = numer * Z + AA[i]`. Should have done that in the first implementation.
Conversely, should we decide to change it, this should get a comment that it is a reformulation of the algorithm from Humlíçek and Schreier.
> 
> I have not done any benchmarking on this; however, do you know if this 20% is fixed or it relative to problem size. That is, there is a good chance the the overhead cost is relatively fixed as all the cases in #11177 are fairly small (the loop is a fixed iteration size). In any case, if performance is a huge concern for this algorithm, then we should consider moving this algorithm into a C extension and/or contributing it directly to `scipy`?

That was just a quick test, as noted for some 1000 frequency points covering the transition region well into the Lorentz wings. Would be a complex investigation to find where and how much this will impact performance, which @fSchreier had all done when first raising this issue.
Very likely this is not a huge concern anywhere, and the pure `scipy` implementation, which can actually be somewhat faster for many of these cases, is already available through the `wofz` option. But one original motivation for this implementation was to have a fast method available even where scipy is not.
So that decision may be revised with the ongoing discussion about making scipy a required dependency, but my main point is that it should not belong into a a code-reformatting PR. I and others already started to transgress when adding f-string changes etc., but this would take it to yet another level IMO.

Reply:
Can't ``numer`` be expanded to like ``AA[i]*Z**j + ...``

Reply:
At least that removes the paren explosi

Reply:
But will explode the number of multiplications I think (at least if I understand what you mean).

Reply:
Or ``sum(a * Z**I for i, a in enumerate(AA))``

Reply:
I'm just going to revert the change and let the original stand. IMOP after reviewing #11177, we should deprecate this algorithm ultimately in favor of the `scipy.special.wofz` algorithm. That however is a discussion for a different day.

Reply:
> Should be the same number, since multiplication is distributive.
> 
> `numer = AA[15]*Z**15 + AA[14]*Z**14 + ...`

Yes, but it performs a total of 15 + 14 + ... = 105 multiplications (assuming it is not taking  recourse to `pow`, which it probably should not for integer powers) as opposed to 15 in the recursion. That actually quite significantly increases execution time, up to a factor of 2.
The version with loop does not introduce additional multiplications, just some extra overhead from reassigning variables.

Reply:
> I'm just going to revert the change and let the original stand. IMOP after reviewing #11177, we should deprecate this algorithm ultimately in favor of the `scipy.special.wofz` algorithm. That however is a discussion for a different day.

Thanks, it's hopefully a bit clearer now. The latter of course will require #12448 to find its way into creating an APE, having it passed and implemented first ;-).
Then the small performance gain will indeed hardly set off the small loss in accuracy, at least not in the Python version. If that last bit of tuning is really needed, might try to implement it in a compiled module, but that would rather be a better task for the Scipy side.